### DISCUSSION TITLE
nlohmann parser crashes intermittently while parsing json string.

### DISCUSSION THREAD

**Discussion by arielbodylotion** [nlohmann/json#2571] [Q&A] [answered]
I am using the nlohmann parser to parse a json string, but my code is crashing intermittently. 

```
const std::string& dataStr; //I receive dataStr as a parameter, this is just to show type of dataStr
std::string jsonString(dataStr.c_str(), dataStr.c_str() + dataStr.length());
nlohmann::json  jsonData = nlohmann::json::parse(jsonString);
```

I don't have the particular json for which the parser is crashing, but I have the crash logs.

`~\Imports\Open\nlohmannjson\include\nlohmann\json.hpp line 11619, parse error - unexpected .`

As per my understanding the above error means the json string received is not valid json. Is my assertion correct? And if yes can it be because the string size is large(~20K length). Any help appreciated, thanks in advance.

**Comment by gregmarr** [comment]
Why are you copying `dataStr` and particularly, why using that form?
I don't see either "unexpected" with "found" or "invalid arg" in the source.  What version of the library are you using?

**Reply by arielbodylotion** [reply]
I am using version 2.1.1 of the parser, and dataStr contains the json string so I am using it to create jsonData.

**Reply by arielbodylotion** [reply]
@gregmarr 
This is the call stack on crash.
```
01  VCRUNTIME140!_CxxThrowException+0x66 [d:\src\vctools\crt\vcruntime\src\eh\throw.cpp @ 74]
02  !nlohmann::basic_json<std::map,std::vector,std::basic_string<char,std::char_traits<char>,std::allocator<char> >,bool,__int64,unsigned __int64,double,std::allocator,nlohmann::adl_serializer>::parser::unexpect+0x116 [nlohmannjson\include\nlohmann\json.hpp @ 11619]
03  !nlohmann::basic_json<std::map,std::vector,std::basic_string<char,std::char_traits<char>,std::allocator<char> >,bool,__int64,unsigned __int64,double,std::allocator,nlohmann::adl_serializer>::parser::parse_internal+0x387 [nlohmannjson\include\nlohmann\json.hpp @ 11584]
04  !nlohmann::basic_json<std::map,std::vector,std::basic_string<char,std::char_traits<char>,std::allocator<char> >,bool,__int64,unsigned __int64,double,std::allocator,nlohmann::adl_serializer>::parser::parse+0x2b [nlohmannjson\include\nlohmann\json.hpp @ 11387]
05  !nlohmann::basic_json<std::map,std::vector,std::basic_string<char,std::char_traits<char>,std::allocator<char> >,bool,__int64,unsigned __int64,double,std::allocator,nlohmann::adl_serializer>::parse<std::_String_const_iterator<std::_String_val<std::_Simple_types<char> > >,0>+0x7d [\nlohmannjson\include\nlohmann\json.hpp @ 6436]
06  !nlohmann::basic_json<std::map,std::vector,std::basic_string<char,std::char_traits<char>,std::allocator<char> >,bool,__int64,unsigned __int64,double,std::allocator,nlohmann::adl_serializer>::parse<std::basic_string<char,std::char_traits<char>,std::allocator<char> >,0>+0x55 [nlohmannjson\include\nlohmann\*json.hpp* @ 6488]
```

**Reply by arielbodylotion** [reply]
I have updated the error as well, please have a look. @gregmarr

**Reply by nlohmann** [reply]
How does the error message look like?

**Reply by gregmarr** [reply]
You can pass `dataStr` directly to `parse()`, you don't need to copy it first.

In this version of the code, this error happens when there is extra text after what it parses as valid input.

               expect(lexer::token_type::end_of_input);

The version you are using is almost 4 years old (February 2017).  I would suggest trying again with the latest version as it has much better error reporting.

**Reply by arielbodylotion** [reply]
ok, thank you.

**Reply by arielbodylotion** [reply]
@nlohmann "parse error - unexpected"

**Reply by nlohmann** [reply]
There should be more after this.

**Reply by arielbodylotion** [reply]
I don't have the error message, I just have the stack trace. The message I gave you was from the last line that was executed in the stack trace. sorry.

**Reply by nlohmann** [reply]
The `what()` message of recent versions of the library contains more information, like the last read token, or expected tokens. Without more information (like a concrete code example), it's hard to help here.

**Reply by arielbodylotion** [reply]
Ok cool, I have started handling the exception thrown by the parser. If it fails again, we can understand what is causing this. Thanks a lot for your help.

**Reply by nlohmann** [reply]
@arielbodylotion Do you need further assistance?

**Comment by arielbodylotion** [comment, marked as answer]
Thanks for the help. IT was actually an issue in the JSON.
